pandas/io/common.py

```diff
@@ -176,19 +177,23 @@ def stringify_path(
     Any other object is passed through unchanged, which includes bytes,
     strings, buffers, or anything else that's not even path-like.
     """
-    if hasattr(filepath_or_buffer, "__fspath__"):
-        # https://github.com/python/mypy/issues/1424
-        # error: Item "str" of "Union[str, Path, IO[str]]" has no attribute
-        # "__fspath__"  [union-attr]
-        # error: Item "IO[str]" of "Union[str, Path, IO[str]]" has no attribute
-        # "__fspath__"  [union-attr]
-        # error: Item "str" of "Union[str, Path, IO[bytes]]" has no attribute
-        # "__fspath__"  [union-attr]
-        # error: Item "IO[bytes]" of "Union[str, Path, IO[bytes]]" has no
-        # attribute "__fspath__"  [union-attr]
-        filepath_or_buffer = filepath_or_buffer.__fspath__()  # type: ignore[union-attr]
-    elif isinstance(filepath_or_buffer, pathlib.Path):
-        filepath_or_buffer = str(filepath_or_buffer)
+    if not convert_file_like and is_file_like(filepath_or_buffer):
+        # GH 38125: some fsspec objects implement os.PathLike but have already opened a
+        # file. This prevents opening the file a second time. infer_compression calls
+        # this function with convert_file_like=True to infer the compression.
+        #
+        # Incompatible return value type (got "Union[_PathLike[str], Union[str,
+        # Union[IO[str], RawIOBase, BufferedIOBase, TextIOBase, TextIOWrapper, mmap]]]",
+        # expected "Union[str, Union[IO[str], RawIOBase, BufferedIOBase, TextIOBase,
+        # TextIOWrapper, mmap]]")  [return-value]
+        # Incompatible return value type (got "Union[_PathLike[str], Union[str,
+        # Union[IO[bytes], RawIOBase, BufferedIOBase, TextIOBase, TextIOWrapper,
+        # mmap]]]", expected "Union[str, Union[IO[bytes], RawIOBase, BufferedIOBase,
+        # TextIOBase, TextIOWrapper, mmap]]")  [return-value]
+        return filepath_or_buffer  # type: ignore[return-value]
```

Comment:
can you not add an assert here (e.g. that its a Buffer)?

Reply:
`Buffer` is a generic type, it seems that `isinstance` does work with generics. Unfortunately, we cannot do `not isinstane(filepath_or_buffer, os.PathLike)`. I could add a `cast` but that is probably as good as the ignore statement.